add option for host to remove guests and end session

danielamormocea opened the pull request:
https://github.com/user-attachments/assets/4220e904-30f5-4d3c-9a50-8fdf38b3f528

Show lists of guests in Settings, and click on trash-icon to remove guest.
Click on "End Session" button to end the current session. This action will lead to a page with the artifact.

maettuu commented:
we need to discuss the BE changes - imo once again communication is missing :(

danielamormocea commented:
@maettuu  Hey :) sure - but that's why we have reviews, if anything is unclear, just pinpoint (also not everything is set in stone - the messaging part I can adapt later after websockets are completely done, I just wanted to have something working for now).
 
There are 3 things I fixed - 2 of them were bugs in the websockets (and because I'm testing directly with FE I might have a better picture of what's going on - there was an axios issue, and there was another issue that when I tried to "end-session" it was telling me that "you can't close a websocket that's already been closed". 

The 3rd thing I changed was just to be able to get the names of the guests (so they appear in the FE - as for the moment it was just the id sent). I also consulted the endpoints sheet - I saw there is one endpoint tagged with "Could" (so not sure if we want to do that or not) - but that's just usable for one guest at a time (so if we have 20 guests that means 20 calls for their names). So I adjusted the existing one (similarly to how we do with the song info) - to pass the guest info necessary. Lmk if you have a better idea of how to do this, or if you have any questions :) 

Feel free to ask directly at the line if anything is still unclear

jemaie commented on server/app/service.py:
```diff
@@ -300,18 +302,26 @@ async def establish_ws_connection_to_channel_by_session_id(websocket: WebSocket,
     @staticmethod
     async def end_all_ws_connections_to_channel_by_session_id(session_id: str) -> None:
+        if session_id not in manager.active_connections:
+            return
```
this is taken care of in the new iteration of the ws code

jemaie commented on server/app/service.py:
```diff
@@ -300,18 +302,26 @@ async def establish_ws_connection_to_channel_by_session_id(websocket: WebSocket,
+            try:
+                await websocket.close(code=1000, reason='Session ended')
+            except WebSocketDisconnect:
+                pass
```
not sure if this makes sense though. closing a ws connection should not be able to throw a WSDisconnect error. otherwise, i'd be glad if you could leave this code out in this PR as a precaution, the incoming PR with the ws changes did a lot. if it is later on necessary, we can again introduce it.

jemaie commented on server/app/service.py:
```diff
@@ -300,18 +302,26 @@ async def establish_ws_connection_to_channel_by_session_id(websocket: WebSocket,
         session = await self.get_session(session_id)
         self.verify_host_of_session(host_id, session)
         host.sessions.remove(session.id)
+        await self.repo.set_user(host)
```
we don't have the attribute "sessions" on the user anymore. these changes (320 & 324) can be omitted or discarded when merging.

jemaie commented on server/app/service.py:
```diff
@@ -291,6 +292,7 @@ async def get_most_popular_recommendation(self, session_id: str) -> Song:
     async def establish_ws_connection_to_channel_by_session_id(websocket: WebSocket, session_id: str) -> None:
         session_connections =  manager.active_connections.get(session_id, set())
         session_connections.add(websocket)
+        manager.active_connections[session_id] = session_connections
```
this method doesn't exist anymore in the big WS change PR. this change can be omitted or discarded during merging

danielamormocea commented:
Updated PR, and also used this PR to add the newest FE changes, to adapt for the most recent BE structure.

Additionally, I added some locks in BE to account for concurrent changes to the session by the host or guests (feel free to adapt this later if you find another solution).

One more thing changed in BE was the snake case to camel case conversion (added CamelModel) - this was needed in BE for nested objects (the playlist in the session) and for websockets (to send camelcase to FE).

jemaie reviewed (commented):
Approved BE changes
